src/client.rs

```diff
@@ -182,9 +172,9 @@ use std::usize;
 ///
 /// [module]: index.html
 #[must_use = "futures do nothing unless polled"]
-pub struct Handshake<T, B = Bytes> {
+pub struct Handshake<'a, T, B = Bytes> {
     builder: Builder,
-    inner: WriteAll<T, &'static [u8]>,
+    inner: Pin<Box<dyn Future<Output = io::Result<T>> + 'a>>,
```

Comment:
With this holding a `Box<dyn Future>`, the type is no longer `Send` or `Sync`, and to add those bounds to the trait object would prevent using this with a `!Send` transport type. It's unfortunate, but that's the reason the `WriteAll` type was there initially. Can we still name the future instead of needing a trait object?

Reply:
The problem is that the WriteAll struct is private now. One solution I have in mind is make the handshake fns async. The handshake path has only one asynchronous element and that is `write_all`. We can just await it right there in the `handshake2` function. WDYT?

Reply:
In hyper we store the handshake future in a state enum.

https://github.com/hyperium/hyper/blob/a503f8b3369ff24ce39ccf1a1da0d1d56366dabb/src/proto/h2/client.rs#L34-L37

Making handshake functions async will return an opaque future which we won't be able to store without Boxing. Will we update that area to async functions in hyper 0.13?  From what I understand, Tokio is also moving towards async fns where possible. (recent examples: `poll_accept` vs `accept`, `poll_ref` vs `recv_ref`)

Reply:
@carllerche thoughts here?

Reply:
Is it fine for now to provide a custom Writeall impl within h2?

Reply:
The idea is that APIs that use `write_all` would be written w/ `async fn`. Can this happen here?

Reply:
This would obviously bubble into hyper, would it be acceptable for hyper to also use `async fn` or is there a blocker somewhere?

Reply:
It could, though that would make it more difficult to allow the client `Executor` to be `Send` or `!Send`, since we couldn't name that spawned future...

Reply:
Done 👍